src/pygeom.c

```diff
@@ -0,0 +1,316 @@
+        is_hex = ((wkb[0] == 48) | (wkb[0] == 49));
+    } else {
+        is_hex = 0;
+    }
+
+    /* Create the reader and read the WKB */
+    reader = GEOSWKBReader_create_r(context_handle);
+    if (reader == NULL) {
+        return NULL;
+    }
+    if (is_hex) {
+        geom = GEOSWKBReader_readHEX_r(context_handle, reader, (unsigned char *) wkb, size);
+    } else {
+        geom = GEOSWKBReader_read_r(context_handle, reader, (unsigned char *) wkb, size);
+    }
+    GEOSWKBReader_destroy_r(context_handle, reader);
+    if (geom == NULL) {
+        return NULL;
+    }
+    result = GeometryObject_FromGEOS(type, geom);
+    if (result == NULL) {
+        GEOSGeom_destroy_r(context_handle, geom);
+        PyErr_Format(PyExc_RuntimeError, "Could not instantiate a new Geometry object");
+    }
+    return result;
+}
+
+
+static PyObject *GeometryObject_new(PyTypeObject *type, PyObject *args,
+                                    PyObject *kwds)
+{
+    void *context_handle = geos_context[0];
+    GEOSGeometry *arg;
+    GEOSGeometry *ptr;
+    PyObject *self;
+    PyObject *value;
+
+    if (!PyArg_ParseTuple(args, "O", &value)) {
+        return NULL;
+    }
+
+    if (PyBytes_Check(value)) {
+        return GeometryObject_FromWKB(type, value);
+    }
+    else if (PyUnicode_Check(value)) {
+        return GeometryObject_FromWKT(type, value);
+    }
+    else {
+        PyErr_Format(PyExc_TypeError, "Expected string or bytes, found %s", value->ob_type->tp_name);
+        return NULL;
+    }
+    ptr = GEOSGeom_clone_r(context_handle, arg);
+    if (ptr == NULL) {
+        RAISE_ILLEGAL_GEOS;
+        return NULL;
+    }
+    self = GeometryObject_FromGEOS(type, ptr);
+    return (PyObject *) self;
+}
+
+static PyMethodDef GeometryObject_methods[] = {
+    {"to_wkt", (PyCFunction) GeometryObject_ToWKT, METH_VARARGS | METH_KEYWORDS,
+     "Write the geometry to Well-Known Text (WKT) format"
+    },
+    {"to_wkb", (PyCFunction) GeometryObject_ToWKB, METH_VARARGS | METH_KEYWORDS,
+     "Write the geometry to Well-Known Binary (WKB) format"
+    },
+    {"from_wkt", (PyCFunction) GeometryObject_FromWKT, METH_CLASS | METH_O,
+     "Read the geometry from Well-Known Text (WKT) format"
+    },
+    {"from_wkb", (PyCFunction) GeometryObject_FromWKB, METH_CLASS | METH_O,
+     "Read the geometry from Well-Known Binary (WKB) format"
+    },
+    {NULL}  /* Sentinel */
+};
+
+PyTypeObject GeometryType = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name = "pygeos.ufuncs.GEOSGeometry",
+    .tp_doc = "Geometry type",
+    .tp_basicsize = sizeof(GeometryObject),
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT,
+    .tp_new = GeometryObject_new,
+    .tp_dealloc = (destructor) GeometryObject_dealloc,
+    .tp_members = GeometryObject_members,
+    .tp_methods = GeometryObject_methods,
+    .tp_repr = (reprfunc) GeometryObject_repr,
+};
+
+int
+init_geom_type(PyObject *m)
+{
+    if (PyType_Ready(&GeometryType) < 0)
+        return NULL;
```

Comment:
There is one remaining (newly introduced) build warning I didn't fully solve. It is complaining here that it is returning NULL (a pointer), while I declared it to return `int`. And also that the rest does not return anything.

But maybe I can return here 0 / -1, and do the `return NULL` in the actual `PyInit_ufuncs` function

Reply:
I think you should indeed return 0 / -1 and depending on that, `return NULL` in the `PyInit_ufuncs` (in `pygeos/ufuncs.c`) if necessary